### PR TITLE
abandoned: feat: added max-test-shards and client-details to additional-app-test-apks

### PR DESCRIPTION
**Pull request by asadsalman**
Fixes #

This PR adds the ability to specify `max-test-shards` and `client-details` for tests under `additional-app-test-apks` like so:
```
additional-app-test-apks:
    - test: lib1-test.apk
      max-test-shards: 3
    - test: lib2-test.apk
      max-test-shards: 1
      client-details:
        modulename: lib2
        key2: val2
```

These configurations really improves support for multi-module projects, usually the thing I want to change the most is number of shards used to test a module. It also lets me differentiate between modules by setting key/values under `client-details`.

Added unit tests.
.

## Checklist

- [ ] Documented
- [x] Unit tested
- [ ] Integration tests updated

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/gc/GcAndroidTestMatrix.kt [hunk 1]
```diff
@@ -38,10 +38,15 @@ object GcAndroidTestMatrix {
         obbFiles: Map<String, String>
     ): Testing.Projects.TestMatrices.Create {
         // https://github.com/bootstraponline/studio-google-cloud-testing/blob/203ed2890c27a8078cd1b8f7ae12cf77527f426b/firebase-testing/src/com/google/gct/testing/launcher/CloudTestsLauncher.java#L120
-        val clientInfo = ClientInfo()
-            .setName("Flank")
-            .setClientInfoDetails(args.clientDetails?.toClientInfoDetailList())
-
+        val clientInfo = if (androidTestConfig is AndroidTestConfig.Instrumentation && androidTestConfig.clientDetails.isNotEmpty()) {
+                            ClientInfo()
```
Comment (Sloox) [inline test_runner/src/main/kotlin/ftl/gc/GcAndroidTestMatrix.kt:42]: Strange formatting, do you have ktlint installed?

**Review by Sloox — COMMENTED** [review]
@asadsalman Hey can you sign the CLA bot to continue? If you are having issues with it or cannot sign it for legal reasons, please open an issue and we will implement this change for you.

**Comment by asadsalman** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by asadsalman** [comment]
I wrote up this draft to get early feedback on this approach. If it looks reasonable to you, I'm happy to proceed with fixing tests/linting/doc updates etc. I based my approach on [this PR](https://github.com/Flank/flank/pull/1126). Let me know if I should proceed.

File: test_runner/src/main/kotlin/ftl/gc/GcAndroidTestMatrix.kt (hunk 1 above)
Comment (pawelpasterz) [reply test_runner/src/main/kotlin/ftl/gc/GcAndroidTestMatrix.kt:42]: @asadsalman or you can just run`./gradlew formatKotlin` before commit

**Comment by pawelpasterz** [comment]
cc @jan-gogo

**Comment by Sloox** [comment]
recheck

**Comment by asadsalman** [comment]
recheck

**Review by jan-gogo — APPROVED** [review]
The implementation looks good, thanks! 👍

File: test_runner/src/main/kotlin/ftl/args/ArgsToString.kt [hunk 2]
```diff
@@ -32,6 +32,20 @@ object ArgsToString {
 
     fun apksToString(devices: List<AppTestPair>): String {
         if (devices.isNullOrEmpty()) return ""
-        return NEW_LINE + devices.joinToString(System.lineSeparator()) { (app, test) -> "        - app: $app\n          test: $test" }
+
+        return NEW_LINE + devices.joinToString(System.lineSeparator()) {
+
+            val environmentVars = if (it.environmentVariables.isNotEmpty()) {
+                "\n          environment-variables:\n" +
```
Comment (Sloox) [inline test_runner/src/main/kotlin/ftl/args/ArgsToString.kt:39]: for the `\n` is it possible to use System.lineseperator?

**Review by Sloox — APPROVED** [review]
Minor comment, else it looks fine by me :+1:

**Comment by asadsalman** [comment]
This PR is now implemented by #1947  since a recent refactor introduced merge conflicts and it was easier to reimplement this as a new PR instead of rebasing and including all those new commits in this PR too.

**Pull request closed without merging**